<!doctype html>

<link href="css/style.css" rel="stylesheet">
<head>
	<!-- Global site tag (gtag.js) - Google Analytics -->
	<script async src="https://www.googletagmanager.com/gtag/js?id=UA-130520215-1"></script>
	<script>
		window.dataLayer = window.dataLayer || [];
		function gtag(){dataLayer.push(arguments);}
		gtag('js', new Date());
		gtag('config', 'UA-130520215-1');
	</script>
	
	<meta charset="utf-8">
	<link rel="icon" href="images/icon.png">
        <title>SHAPEIT4</title>
	<meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <link rel="stylesheet" href="css/bootstrap.min.css">
        <link rel="stylesheet" href="css/style.css">
</head>

<body>

<!--Home page style-->
<header id="header" class="header">
	<h1><b>SHAPEIT4</b></h1>
	<h2><b>S</b>egmented <b>HAP</b>lotype <b>E</b>stimation and <b>I</b>mputation <b>T</b>ools version <b>4</b></h2>
</header>

<div class="container">
	<ul class="nav navbar-nav">
        <li><a href="#introduction">Introduction</a></li>
		<li><a href="#version">Versions</a></li>
		<li><a href="#installation">Installation</a></li>
		<li><a href="#documentation">Documentation</a></li>
		<li><a href="#history">History</a></li>
		<li><a href="#contact">Contact</a></li>
	</ul>
</div>

<!-- Section INFORMATION -->
<section id="introduction" class="sections lightbg">
	<div class="container">
		<h1 style="color:red;">Importance Notice</h1>
		<div class="separator"></div>
		<p>Version 5 is now available and contains a series of major updates:</p>
		<ul>
			<li>Multi-threading of PBWT scans and initializing phasing sweep. This reduces elapsed running times.</li>
			<li>Support of duo/trio phasing. Kids are scaffolded at positions that can be directly phasing using Mendel transmission logic.</li>
			<li>Improvements of IBD2 protection.</li>
			<li>New methods for rare variant phasing that dratically speeds up phasing of large WGS datasets.</li>
			<li>Release of multiple utility tools to split the genome in chunks and ligate phased chunks back into whole chromosome haplotypes.</li>
		</ul>
		<div style="text-align:center">
			<p style="color:red;"><b>Version 4 is not supported anymore.</b> Version 5 is available from the following places:</p>
			<font size="+3"><p><a href="https://odelaneau.github.io/shapeit5/">SHAPEIT5 website</a> / <a href="https://github.com/odelaneau/shapeit5">SHAPEIT5 github</a></p></font>
		</div>
	</div>
</section>




<!-- Section INFORMATION -->
<section id="introduction" class="sections lightbg">
	<div class="container">
		<h1>Introduction</h1>
		<div class="separator"></div>
		<p>SHAPEIT4 is a fast and accurate method for <a href="https://en.wikipedia.org/wiki/Haplotype_estimation">estimation of haplotypes</a> (aka phasing) for SNP array and high coverage sequencing data. The version 4 is a refactored and improved version of the SHAPEIT algorithm with multiple key additional features:</p>
		<ul>
			<li>It includes a Positional Burrow Wheeler Transform (PBWT) based approach to quickly select a small set of informative conditioning haplotypes to be used when updating the phase of an individual.</li>
			<li>We have changed that way in which phase information in sequencing reads is input into the model. We now recommend the use of the WhatsHap tool as a pre-processing step to extract phase information from a bam file.</li>
			<li>It accounts for sets of pre-phased genotypes (i.e. haplotype scaffold). The scaffold can be derived either from family data or large reference panels.</li>
			<li>It reads and writes files using HTSlib for better I/O performance in either VCF or BCF formats.</li>
			<li>The genotype graph and HMM routines have been re-implemented for better hardware usage and performance.</li>
			<li>The source code is provided in an open source format (licence MIT) on github.</li>
		</ul>
		<p>If you use the SHAPEIT4 in your research work, please cite the following paper:</p>
		<p><a href="https://www.nature.com/articles/s41467-019-13225-y">Olivier Delaneau, Jean-Francois Zagury, Matthew R Robinson, Jonathan Marchini, Emmanouil Dermitzakis. Accurate, scalable and integrative haplotype estimation. Nat. Comm. 2019.</a></p>
		
		<p>SHAPEIT4 is available under the <a href="https://en.wikipedia.org/wiki/MIT_License">MIT licence</a> on the github repository <a href="https://github.com/odelaneau/shapeit4">https://github.com/odelaneau/shapeit4</a>.</p>
		<p>Previous versions of SHAPEIT are still available on original webpages: <a href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html">SHAPEIT2</a> and <a href="https://jmarchini.org/shapeit3/">SHAPEIT3</a>.</p>
	</div>
</section>


<!-- Section VERSIONS -->
<section id="version" class="sections lightbg">
	<div class="container">
		<h1>Versions [ALWAYS USE LAST VERSION!]</h1>
		<div class="separator"></div>
		
		<p><b>Version 4.2.2</b>: Small update, it contains the following changes compared to previous version:</p>
		<ul>
			<li>Improvements of HMM routines introduced in v4.2.0 are now also working when using scaffold.</li>
		</ul>    
		<p><b>Version 4.2.1</b>: Small update, it contains the following changes compared to previous version:</p>
		<ul>
			<li>Fixing a bug happening when compiling using g++9 / g++10, caused by a function not returning an expected bool return statement.</li>
			<li>Bug was not present when compiling with g++7. Thanks to Carl Nettelblad who spotted the problem.</li>
		</ul>    
		<p><b>Version 4.2.0</b>: Fourth release. This is a major update, it contains the following changes compared to previous version:</p>
		<ul>
			<li>All HMM routines have been rewritten for better performance. The method is now ~2x faster on SNP array data and more on sequencing data.</li>
			<li>The detection of IBD2 tracks has been greatly improved. All options --ibd2-* are now deprecated. </li>
			<li>Multi-threading is now used when compressing/decompressing VCF/BCF files (Thanks to Simone Rubinacci for this!).</li>
		</ul>    
		<p><b>Version 4.1.3</b>: Third release. Contains the following changes compared to previous version:</p>
		<ul>
			<li>Bug correction on phase sets. This bug was introduced by the new engin for sporadic missing data imputation.</li>
			<li>Bug correction for scaffold functionalities. HMM computations are now double precision and not anymore single precision for samples having a scaffold.</li>
		</ul>    
		<p><b>Version 4.1.2</b>: Second release. Contains the following changes compared to previous version:</p>
		<ul>
			<li>Bug correction on the PBWT selection. The previous functions were not working well for data with a high density of variants (e.g. sequencing data). We added an option --sequencing to adjust the default parameters to what we think is adapted to sequencing data.</li>
			<li>Better algorithm for imputing missing data. The previous algorithm was performing badly when the amount of missing data is high (e.g. > 20%). The new one is doing much better in such particular cases.</li>
			<li>We would like to thank <b>Vivek Appadurai</b> and <b>Andrew J. Schork</b> (Institute of Biological Psychiatry, Mental Health Center Sct. Hans, Roskilde, Denmark) for their very valuable inputs that made these improvments possible.</li>
		</ul>    
		<p><b>Version 4.1.1</b>: Contains the following changes compared to previous version:</p>
		<ul>
			<li>Bug correction on the detection of IBD2 segments. The previous function was not working in some cases involving increased switch errors rates for duplicates, twins and siblings.</li>
			<li>Better genetic map interpolation, i.e. way faster.</li>
			<li>All windowing/chunking is now done on the basis on cM distances and not anymore on Mb distances (option --window and --pbwt-modulo). This improves the accuracy for low density SNP arrays.</li>
			<li>Addition of two new options to control the PBWT array storage (<code>--pbwt-mac</code> and <code>--pbwt-mdr</code>). See documentation. 
			<li>Usage of single-precision floating-point numbers and not anymore double-precisions. This increases speed.</li>
			<li>Explicit usage of AVX2 capabilities of modern CPUs. The most computationally demanding HMM routines have been re-written using AVX2 intel intrinsincs functions. This increases speed.</li>  
		</ul>    
		<p><b>Version 4.0</b>: First release.</p>
	</div>
</section>

<!-- Section INSTALLATION -->
<section id="installation" class="sections">
    <div class="container">
        <div class="scroll-top"><a href="#header"><div class="scrollup">&uarr;</div></a>
        <div class="heading text-center">
            <h1>installation</h1>
            <div class="separator"></div>
        </div>
        <div class="row">
        <p><b>Important notes before starting:</b></p>
        <ol>
			<li>Compile SHAPEIT4 on the same machine it will be run if possible. This significantly increases performance.</li>
			<li>Run preferentially SHAPEIT4 on CPUs supporting AVX2 instruction set as the code has been heavily optimized in that direction. AVX2 is usually available for CPUs from 2012+. Run <code>cat /proc/cpuinfo</code> in order to check.</li>
			<li>Compile SHAPEIT4 using g++ version >= 5 at the very least. I personally use version v7.4.0. Compiling it using g++ v4 does not work. Run <code>g++ --version</code> in order to check. I've never tried to compile it using Intel compiler.</li>
        </ol>
                
		<p>Make sure that the two following libraries are installed on your system:</p>
		<ol>
			<li><a href="http://www.htslib.org/">HTSlib</a>: A great C library for reading/writing high-throughput sequencing data.</li>
			<li><a href="https://www.boost.org/">BOOST</a>: A free peer-reviewed portable C++ source libraries. SHAPEIT4 uses two specific BOOST libraries: iostreams and program_options.</li>
		</ol>
		<p>Make sure that the following standard library flags can be used by g++ on your system:</p>
		<ul>
			<li><code>-lz</code>, <code>-lbz2</code> and <code>-llzma</code> for reading/writing compressed files.</li>
			<li><code>-lm</code> for basic math operations.</li>
			<li><code>-lpthread</code> for multi-threading.</li>
		</ul>
		<p>You can do so by checking the outcome of the following commands:</p>
		<code>locate -b '\libz.so'</code><br>
		<code>locate -b '\libbz2.so'</code><br>
		<code>locate -b '\liblzma.so'</code><br>
		<code>locate -b '\libm.so'</code><br>
		<code>locate -b '\libpthread.so'</code><br>
		<p>Then, download the last stable version of the SHAPEIT4 code here: <a href="https://github.com/odelaneau/shapeit4/releases/">RELEASES</a></p>
		<p>Navigate to the downloaded folder using <code>cd shapeit4</code>.</p>
		<p>You'll find there the following folders and files:</p>
		<ul>
			<li><code>bin</code>: folder for the compiled binary.</li>
			<li><code>docs</code>: source code of this website.</li>
			<li><code>maps</code>: genetic maps for human taken from HapMap.</li>
			<li><code>obj</code>: folder with all binary objects.</li>
			<li><code>src</code>: folder with all SHAPIT4 source code.</li>
			<li><code>test</code>: folder with small example datasets.</li>
			<li><code>README.md</code>: front page of the GitHub.</li>
			<li><code>LICENSE</code>: copy of the MIT license.</li>
			<li><code>makefile</code> to compile the code.</li>
		</ul>
		<p>Edit the <code>makefile</code> at lines 5-6 and 9-11 so that the following variables are correctly set up (look at the paths already there for an example):</p>
		<ul>
			<li><code>HTSLIB_INC</code> (line 5): path to the HTSlib header files,</li>
			<li><code>HTSLIB_LIB</code> (line 6): path to the static HTSlib library (file <code>libhts.a</code>),</li>
			<li><code>BOOST_INC</code> (line 9): path to the BOOST header files (often <code>/usr/include</code>),</li>
			<li><code>BOOST_LIB_IO</code> (line 10): path to the static BOOST <i>iostreams</i> library (file <code>libboost_iostreams.a</code>),</li>
			<li><code>BOOST_LIB_PO</code> (line 11): path to the static BOOST <i>program_options</i> library (file <code>libboost_iostreams.a</code>),</li>
		</ul>
		<p>The static libraries (<i>*.a</i> files) can be located with this command:</p>
		<code>locate libboost_program_options.a libboost_iostreams.a libhts.a</code>
		<p>Once all paths correctly set up, proceed with the SHAPEIT4 compilation using <code>make</code>.</p>
	</div>
    </div> <!-- /container -->       
</section>

<!-- Section DOCUMENTATION -->
<section id="documentation" class="sections lightbg">
    <div class="container">
        <div class="scroll-top"><a href="#header"><div class="scrollup">&uarr;</div></a>
        <div class="heading text-center">
            <h1>documentation</h1>
            <div class="separator"></div>
        </div>
        <div class="row">
		<h2>1. Simple run</h2>
		<p>To do so with default parameters, use the following command line:</p>
		<code>shapeit4 --input unphased.vcf.gz --map chr20.b37.gmap.gz --region 20 --output phased.vcf.gz</code>
		<p>All four options are mandatory and their descriptions are:</p>
		<ul>
			<li><code>--input</code> specifies the input file containing unphased genotype data.</li>
			<li><code>--map</code> specifies the genetic maps. Maps for humans can be found <a href="https://github.com/odelaneau/shapeit4/tree/master/maps">HERE</a>.</li>
			<li><code>--region</code> specifies the genomic region to be phased.</li>
			<li><code>--output</code> specifies the output VCF/BCF file containing the phased genotype data (i.e. haplotypes).</li>
		</ul>
		<p>The file provided to <code>--input</code> needs to be indexed. This can be done using <code>bcftools index unphased.vcf.gz</code>.</p>

		<h2>2. Phasing a chunk of data</h2>
		<p>To phase the 1Mb of input data located in the genomic interval 2Mb-3Mb, use:</p>
		<code>shapeit4 --input unphased.vcf.gz --map chr20.b37.gmap.gz --region <b>20:2000000-3000000</b> --output phased.vcf.gz</code>
		<p>Of note, the <code>--region</code> option is mandatory. Double check that the chromosome ID matches one of those specified in the VCF file. A common mistake is to use <i>chr20</i> for chromosome 20 for example, while it is specified as <i>20</i> in the VCF file.</p>

		<h2>3. BCF files</h2>
		<p>SHAPEIT4 file automatically detects the format of the input file. To tell it to use BCF as output file, just use the correct filename extension as follows:</p>
		<code>shapeit4 --input unphased.<b>bcf</b> --map chr20.b37.gmap.gz --region 20 --output phased.<b>bcf</b></code>

		<h2>4. Multi-threading</h2>
		<p>To phase the example dataset on 4 CPU cores using 4 threads, use the <code>--thread</code> option as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--thread 4</b></code>

		<h2>5. Iteration scheme</h2>
		<p>To specify your own iteration sequence, use the <code>--mcmc-iterations</code> option:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--mcmc-iterations 6b,1p,2b,1p,2b,8m</b></code>
		<p>The three different kinds of iterations are:</p>
		<ul>
			<li><code>b</code>: <i>burn-in iteration</i>. Only haplotype sampling and no storage of transition probabilities. To be used for convergence purpose.</li>
			<li><code>p</code>: <i>pruning iteration</i>. New haplotype are sampled and transition probabilities are used to trim unlikely paths in the genotype graphs. Using few of them once the algotrithm has converged helps to get better estimates. They have to be used after few burn-in iterations and before main iterations. It is recommended to run them paired with burn-in iterations, i.e. <code>1b,1p</code>.</li>
			<li><code>m</code>: <i>main iteration</i>. New haplotype are sampled and transition probabilities are stored and averaged in memory. To be used as final iteration stage to produce final estimates.</li>
		</ul>
		<p>By default, the sequence used by SHAPEIT4 is <code>5b,1p,1b,1p,1b,1p,5m</code>. It provided a good trade-off between speed and accuracy in our experiments. If running time is not an issue, you may consider increasing the number of iterations using <code>--mcmc-iterations 10b,1p,1b,1p,1b,1p,1b,1p,10m</code> for instance.</p>
		
		<h2>6. Tuning the PBWT based selection [v4.1.0 and above]</h2>
		<p>Reducing the number of conditioning neighbours in the PBWT can be achieved using the <code>--pbwt-depth</code> option. The default value is 4. Decreasing it results in faster runs at the cost of some accuracy. Conversely, increasing it to 8 for instance can lead to better accuracy:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--pbwt-depth 8</b></code>
		<p>Changing variants at which PBWT indexes are stored can be done using <code>--pbwt-modulo</code>. The way this option works differs from v4.0. Now variants at which PWT arrays are stored depend on cM distances between them. This distance is given by <code>--pbwt-modulo</code>. By default, PBWT storage happens every 0.02cM. Decreasing this distance leads to higher memory usage. </p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--pbwt-modulo 0.005</b></code>
		<p>Some variants can also be ignored when storing PBWT indexes based on (i) their Minor Allele Count (<code>--pbwt-mac</code>) and/or (ii) Missing Data Rate (<code>--pbwt-mdr</code>). Defaults are (i) MAC>=2 and (ii) MDR <= 0.5. This can be changed as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--pbwt-mac 5 --pbwt-mdr 0.2</b></code>
		<p>You can also change the size of the genomic window into which phasing is carried out using the option <code>--window</code>. The way this option works differs from v4.0. Now the size of the window is based n cM and not anymore on Mb. The default value is 2.5cM. Increasing it results in more conditioning haplotypes being used and therefore increased running times. It is preferable to use <code>--pbwt-depth</code> instead of <code>--window</code> to increase the number of conditioning haplotypes in order to increase accuracy. Anyway, this can be done as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--window 5</b></code>
		
		<h2>7. IBD2 blacklisting [Deprecated in v4.2.0 and above]</h2>
		<p>Long tracks IBD2 shared between pairs of samples can seriously harm accuracy, which typically occurs in the cases of siblings or in more extreme cases, twins or duplicates.
		SHAPEIT4 implements an automatic screening to detect long runs of genotype matching between pairs of samples, so that it can account for it during the phasing procedure.
		In practice, assuming two individuals I1 and I2 to be IBD2 in a given region, SHAPEIT4 will never use haplotypes from I2 to phase I1 in this region nor use haplotypes from I1 to phase I2.
		This procedure, that we call IBD2 blacklisting, can be tuned with the options <code>--ibd2-*</code>. </p>
		<p>You can first specify the minimal length of IBD2 tracks to be considered in the blacklisting as shown below for 5cM tracks:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--ibd2-length 5</b></code>
		<p>You can also control the set of variants to be considered in the IBD2 matching in terms of MAF (ex. 5%), missing data rate (ex. 20%) and mininal number of SNPs (ex. 500):</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--ibd2-maf 0.05 --ibd2-mdr 0.2 --ibd2-count 500</b></code>
		<p>Finally, if you want to have a view of the blacklisting done, use <code>--idb2-output</code> as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--ibd2-output myIBD2blacklist.txt.gz</b></code>
		
		<h2>8. Sequencing data or SNP array data?</h2>
		<p>Most of the default parameters are set up to work well in case of SNP array data (low variant density). In case of sequencing data, some adjustements are required, notably for the PBWT array storage (<code>--pbwt-*</code>).
		See the relevant sections in the documentation to know more about them. For simplicity, we now have the option <code>--sequencing</code> to adjust the default parameters for sequencing data. Example:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--sequencing</b></code>    
			
		<h2>9. Specifying initial estimates</h2>
		<p>By default, SHAPEIT4 uses a quick PBWT based algorithm to initialize the haplotypes. This can be disabled using <code>--pbwt-disable-init</code>. This is particularly useful if you want to specify your own starting point for the MCMC or if you want to refine some haplotypes you estimated in the past. Initial haplotype estimates are specified in the input VCF/BCF using the option <code>--input</code> and phased genotypes, e.g. <code>0|1</code>. Example command line:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--pbwt-disable-init</b></code>
		<p>Of note, this option is not available in v4.1</p>

		<h2>10. Reproducibility</h2>
		<p>Making reproducible runs can sometimes be useful. To do so, you need to specify the random generator seed using <code>--seed</code> and to use <b>a single thread</b>. Using multi-threading prevents reproducibility.</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--seed 123456</b></code>

		<h2>11. Using a reference panel</h2>
		<p>Taking into account reference haplotypes when phasing can improve the accuracy of the estimates, especially when the number of reference haplotypes is much large than the number of individuals to be phased. To do so, proceed as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 <b>--reference reference.bcf</b> --output phased.bcf</code>
		<p>SHAPEIT4 only retain variants that are in the overlap (i.e. intersection) between the two panels specified with <code>--input</code> and <code>--reference</code>. In practice, it proceeds exactly the same way than <code>bcftool isec -c none</code> and therefore keeps only variants with chromosome ID, position, REF and ALT alleles that perfectly match between the two panels.</p>
		<p>The file specified  by <code>--reference</code> needs to be indexed. This can be done using <code>bcftools index reference.bcf</code>.</p>		

		<h2>12. Using haplotype scaffolds</h2>
		<p>Haplotype scaffolds can be used to incorporate prior information about the phase of some heterozygous genotypes, coming from large reference panels of haplotypes or family information for instance. A scaffold is usually defined at a sparser set of variants than the main panel. Example run:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 <b>--scaffold scaffold.bcf</b> --output phased.bcf</code>
		<p>Similarly to <code>--reference</code>, only variants that are in the intersection with the main panel are retained. Variants in the scaffold and the main panels are not phased and remain unchnaged. Variants not in the scaffold but in the main panel are phased onto the scaffold. Intersection is done similarly to <code>bcftool isec -c none</code>. Only individuals that appear in the two files are scaffolded (IDs need to match), those that are in the main panel and not in the scaffold are not scaffolded but appear in the output. Those that are in the scaffold and not in the main panel are not used and do not appear in the output.</p>
		<p>The file specified  by <code>--scaffold</code> needs to be indexed. This can be done using <code>bcftools index scaffold.bcf</code>.</p>
		
		<h2>13. Using phase sets</h2>
		<p>Phase sets are groups of heterozygous genotypes at which phase have been infered using haplotype assembly. To achieve haplotype assembly, we recommend using <a href="https://whatshap.readthedocs.io/en/latest/">WhatsHap</a> but other software can also be considered. Phase sets are specified in the VCF/BCF file format using the <code>PS</code> field. Note that the <code>PS</code> needs to be defined in the VCF/BCF as a 32bits integer since we read them using the C function <code>bcf_get_format_int32</code>. If you define them as string, they won't be taken into account. Example run:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 <b>--use-PS 0.0001</b> --output phased.bcf</code>
		<p>The argument of the <code>--use-PS</code> defines the expected error rate in the phase sets. You need to change the value depending on your confidence in the haplotype assembly calls. In most of the cases, we recomment using <i>0.0001</i>.</p>

 		<h2>14. Complex run</h2>
		<p>SHAPEIT4 allows using all combinations of options described above, such as:</p>
		<code>shapeit4 --input unphased.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--map chr20.b37.gmap.gz \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--region 20 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--reference reference.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--scaffold scaffold.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--use-PS 0.0001 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--thread 4 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--log phased.log\<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--output phased.bcf<br></code>
		<p>This command allows phasing the example dataset using (i) a reference panel, (ii) a scaffold of haplotypes and (iii) the phase sets specified in the input file, all this in a multi-threaded manner.</p>

		<h2>15. Log file</h2>
		<p>To record all the verbose that appear on the screen, use the <code>--log</code> option as follows:</p>
		<code>shapeit4 --input unphased.bcf --map chr20.b37.gmap.gz --region 20 --output phased.bcf <b>--log phased.log</b></code>
		<p>We strongly recommend to use this option for any run.</p>

		<h2>16. Option summary</h2>
		<p>The full list of options can be obtained by running the command:</p>
		<code>shapeit4 <b>--help</b></code>
		<p>This should output this list of options:</b>
		 <table style="width:100%" border="1">
			<tr>
				<th width="20%">Long form</th>
				<th width="15%">Short form</th>
				<th width="10%">Argument</th>
				<th width="55%">Description</th>
			</tr>
			<tr>
				<td><code>--help</code></td>
				<td>NA</td>
				<td>NA</td>
				<td>Brief description for the options.</td>
			</tr>
			<tr>
				<td><code>--seed</code></td>
				<td>NA</td>
				<td>INT</td>
				<td>Seed for random number generator. Default is 15052011.</td>
			</tr>
			<tr>
				<td><code>--thread</code></td>
				<td><code>-T</code></td>
				<td>INT</td>
				<td>Seed for random number generator. Default is 15052011.</td>
			</tr>
			<tr>
				<td><code>--input</code></td>
				<td><code>-I</code></td>
				<td>STRING</td>
				<td>Input VCF/BCF file to be phased.</td>
			</tr>
			<tr>
				<td><code>--reference</code></td>
				<td><code>-H</code></td>
				<td>STRING</td>
				<td>Reference haplotypes in VCF/BCF.</td>
			</tr>
			<tr>
				<td><code>--scaffold</code></td>
				<td><code>-S</code></td>
				<td>STRING</td>
				<td>Haplotype scaffold in VCF/BCF.</td>
			</tr>
			<tr>
				<td><code>--map</code></td>
				<td><code>-M</code></td>
				<td>STRING</td>
				<td>Genetic map. Mandatory in v4.0. In v4.1, constant rate of 1cM per Mb is assumed when the map is omitted.</td>
			</tr>
			<tr>
				<td><code>--region</code></td>
				<td><code>-R</code></td>
				<td>STRING</td>
				<td>Target region (e.g. chr20:1000000-2000000).</td>
			</tr>
			<tr>
				<td><code>--use-PS</code></td>
				<td>NA</td>
				<td>FLOAT</td>
				<td>Use phase sets in input files assuming an error rate of FLOAT.</td>
			</tr>
			<tr>
				<td><code>--mcmc-iterations</code></td>
				<td>NA</td>
				<td>STRING</td>
				<td>Iteration scheme for the MCMC. Default is 5b,1p,1b,1p,1b,1p,5m.</td>
			</tr>
			<tr>
				<td><code>--pbwt-disable-init</code></td>
				<td>NA</td>
				<td>NA</td>
				<td>Do not initialise haplotypes by PBWT (rephase input haplotype data). Only available in v4.0.</td>
			</tr>
			<tr>
				<td><code>--pbwt-depth</code></td>
				<td>NA</td>
				<td>INT</td>
				<td>Depth of PBWT indexes to condition on. Default is 4.</td>
			</tr>
			<tr>
				<td><code>--pbwt-modulo</code></td>
				<td>NA</td>
				<td>INT</td>
				<td><b>v4.0:</b> Storage frequency of PBWT indexes in variant numbers. Default is 8 (i.e. storage every 8 variants). <b>v4.1:</b> Storage frequency of PBWT indexes in cM. Default is 0.025 (i.e. storage every 0.025 cM). Use a low value for sequencing data and the default value for SNP array data. </td>
			</tr>
			<tr>
				<td><code>--pbwt-modulo</code></td>
				<td>NA</td>
				<td>INT</td>
				<td><b>v4.1:</b> Default value for Storage frequency of PBWT indexes in cM for SNP array data (storage every 0.02 cM).</td>
			</tr>
			<tr>
				<td><code>--pbwt-mac</code></td>
				<td>NA</td>
				<td>INT</td>
				<td><b>v4.1</b>. Storage of PBWT arrays only happens at variants with this Minor Allele Count. Default is 2, this excludes singletons and monomorphic sites. </td>
			</tr>
			<tr>
				<td><code>--pbwt-mdr</code></td>
				<td>NA</td>
				<td>FLOAT</td>
				<td><b>v4.1</b>. Storage of PBWT arrays only happens at variants with this Missing Data Rate. Default is 0.05 (5%), this excludes sites with more than 5% missing data. </td>
			</tr>
			<tr>
				<td><code>--sequencing</code></td>
				<td>NA</td>
				<td>NA</td>
				<td><b>v4.1:</b> Set up better default settings for sequencing data (--pbwt_modulo 0.0005 --ibd2-maf 0.0001 --ibd2-count 10000).</td>
			</tr>
			<tr>
				<td><code>--window</code></td>
				<td>NA</td>
				<td>FLOAT</td>
				<td>Minimal size of the phasing window. <b>v4.0:</b> Default is 2Mb (i.e. 2e6). <b>v4.1:</b> Default is 2.5cM.</td>
			</tr>
			<tr>
				<td><code>--output</code></td>
				<td><code>-O</code></td>
				<td>STRING</td>
				<td>Phased haplotypes in VCF/BCF format.</td>
			</tr>
			<tr>
				<td><code>--log</code></td>
				<td>NA</td>
				<td>STRING</td>
				<td>Log file.</td>
			</tr>
		</table> 
		<p>Using short form options, the complex command line described above (section 12) is:</p>
		<code>shapeit4 -I unphased.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-M chr20.b37.gmap.gz \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-R 20 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-H reference.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-S scaffold.bcf \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--use-PS 0.0001 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-T 4 \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;--log phased.log \<br>
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-O phased.bcf<br></code>
                      
		<h2>15. Screen output decrypted</h2>
		<img src="images/verbose.png">

        </div>
    </div> <!-- /container -->       
</section>

<!-- Section HISTORY -->
<section id="history" class="sections">
    <div class="container">
        <div class="scroll-top"><a href="#header"><div class="scrollup">&uarr;</div></a>
        <div class="heading text-center">
            <h1>DEvelopment History</h1>
            <div class="separator"></div>
        </div>
        <div class="row">
		<p>SHAPEIT has been developed by multiple researchers over the years:</p>
		<ul>
			<li>Olivier Delaneau (<b>OD</b>), University of Lausanne, Switzerland. OD is the main developer of the code from version 1 to 4 and has supervised the development of version 4.</li>
			<li>Jonathan Marchini (<b>JM</b>), University of Oxford, UK. JM has supervised the development of versions 2 and 3.</li>
			<li>Jean-François Zagury (<b>JFZ</b>), Conservatoire des arts et métiers, France. JFZ has supervised the development of version 1.</li>
			<li>Jared O’Connell (<b>JO</b>), University of Oxford, UK. JO has co-developed version 3 and extended it to process closely related samples.</li>
		</ul>
		
		<p>SHAPEIT has been developed across the following institutions:</p>
		<table style="width:100%" border="0" cellpadding="10px">
			<tr>
				<th width="20%"><img src="images/cnam.jpg"></th>
				<th width="20%"><img src="images/peptinov.jpg"></th>
				<th width="20%"><img src="images/oxford.png"></th>
				<th width="20%"><img src="images/geneva.png"></th>
				<th width="20%"><img src="images/lausanne.jpg"></th>
			</tr>
		</table>

		<p>In 2011, version 1 has been released. This version introduced the compact graph representation of the sampling space so that Hidden Markov Model computations can be efficiently done [1].</p>
		<p>In 2012, version 2 has been released. This version included the approach introduced in Impute2 to select subsets of conditioning haplotypes by Hamming distance minimization and therefore to speed up computations [2]. </p>
		<p>In 2013, version 2 has been extended to leverage phase information contained in sequencing reads [3].</p>
		<p>In 2013, version 2 has been extended to process datasets with closely related individuals [4]. </p>
		<p>In 2014, version 2 has been extended to handle genotype likelihoods and use haplotype scaffolds in the frame of the 1000 Genomes project [5]. </p>
		<p>In 2016, version 3 has been released. This version was designed to process biobank scale datasets in the frame of the UK Biobank study [6].</p>
		<p>In 2018, version 4 has been released. This version was completely refactored, includes most previous extensions into a unified framework and is open source [7]. </p>

		<p><i>[1] O. Delaneau, J. Marchini, JF. Zagury (2012) A linear complexity phasing method for thousands of genomes. Nat Methods. 9(2):179-81. </i></p>
		<p><i>[2] O. Delaneau, JF. Zagury, J. Marchini (2013) Improved whole chromosome phasing for disease and population genetic studies. Nat Methods. 10(1):5-6.  </i></p>
		<p><i>[3] O. Delaneau, B. Howie, AJ. Cox, JF. Zagury JF, J. Marchini. Haplotype estimation using sequencing reads. Am J Hum Genet. 2013 Oct 3;93(4):687-96. </i></p>
		<p><i>[4] J. O'Connell, D. Gurdasani, O. Delaneau, N. Pirastu, S. Ulivi, M. Cocca, M. Traglia, J. Huang, JE. Huffman, I. Rudan, R. McQuillan, RM. Fraser, H. Campbell, O. Polasek, G. Asiki, K. Ekoru, C. Hayward, AF. Wright, V. Vitart, P. Navarro, JF. Zagury, JF Wilson, D. Toniolo, P. Gasparini, N. Soranzo, MS. Sandhu, J. Marchini. A general approach for haplotype phasing across the full spectrum of relatedness. PLoS Genet. 2014 Apr 17;10(4):e1004234. </i></p>
		<p><i>[5] O. Delaneau, J. Marchini; 1000 Genomes Project Consortium. Integrating sequence and array data to create an improved 1000 Genomes Project haplotype reference panel. Nat Commun. 2014 Jun 13;5:3934. </i></p>
		<p><i>[6] J. O’Connell, K. Sharp, N. Shrine, L. Wain, I. Hall, M. Tobin, JF. Zagury, O. Delaneau, J. Marchini. Haplotype estimation for biobank scale datasets. Nat Genet. 2016 Jul; 48(7): 817–820. </i></p>
		<p><i>[7] O. Delaneau, JF. Zagury, M. Robinson, J. Marchini, ET. Dermitzakis. Integrative haplotype estimation with sub-linear complexity. BioRxiv 2018.</i></p>
	</div>
    </div> <!-- /container -->       
</section>

<section id="contact" class="sections lightbg">
    <div class="container">
        <div class="scroll-top"><a href="#header"><div class="scrollup">&uarr;</div></a>
        <div class="heading text-center">
            <h1>Contact</h1>
            <div class="separator"></div>
        </div>
        <div class="row text-center">
		<p>Prof. <b>Olivier Delaneau</b><br>
		Department of Computational Biology<br>
		University of Lausanne<br>
		olivier.delaneau@gmail.com<br>
		+41 21 692 40 55<br>
		<a href="https://odelaneau.github.io/lap-page/">https://odelaneau.github.io/lap-page/</a></p>
	</div>
    </div> <!-- /container -->       
</section>



<!--Footer-->
<footer id="footer" class="footer">
    <div class="container">
        <div class="row">
            <div class="col-md-6 col-sm-6 col-xs-12">
                <div class="copyright">
                    <p>&copy; 2018 All Rights Reserved</p>
                </div>
            </div>
        </div>
    </div>
</footer>

</body>

